File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);

        for (int i = 0; i < nums.length; i++){
            int start = i + 1, end = nums.length - 1;

            if(i > 0 && nums[i] == nums[i - 1]) continue;

            while (start < end){
                int sum = nums[start] + nums[end];
                if(sum == -nums[i]){
                    List<Integer> lst = new ArrayList<>(Arrays.asList(nums[i], nums[start], nums[end]));
                    result.add(lst);
                    start++;
                    end--;
                    while(start < end && nums[start] == nums[start - 1]) start++;
                    while(start < end && nums[end] == nums[end + 1]) end--; 
                }
                else if(sum < -nums[i]) start++;
                else end--;
            }
        }

        return result;
    }
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        ans: list = []
        n: int = len(nums)
        for i in range(0, n):
            left: int = i+1
            right: int = n - 1

            if i > 0 and nums[i] == nums[i - 1]: continue

            while left < right:
                if nums[left] + nums[right] == -nums[i]:
                    ans.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while nums[left] == nums[left - 1] and left < right: left += 1
                    while nums[right] == nums[right + 1] and left < right: right -= 1
                elif nums[left] + nums[right] > -nums[i]: right -= 1
                else: left += 1
                
        
        return ans